src/common/snippets/src/pass/propagate_precision.cpp

```diff
@@ -0,0 +1,298 @@
+            }
+
+            for (size_t i = 0; i < op->get_output_size(); ++i) {
+                auto output = op->output(i);
+
+                if (output.get_element_type() != op_output_types[i]) {
+                    was_updated = true;
+                    auto convert = std::make_shared<ngraph::snippets::op::ConvertSaturation>(
+                        output,
+                        op_output_types[i]);
+                    ngraph::copy_runtime_info(output.get_node_shared_ptr(), convert);
+
+                    for (auto& input : output.get_target_inputs()) {
+                        auto child = input.get_node();
+                        if (child == convert.get()) {
+                            continue;
+                        }
+
+                        input.replace_source_output(convert->output(0));
+
+
+                        if (ngraph::is_type<ngraph::op::Result>(input.get_node())) {
+                            input.get_tensor_ptr()->add_names(output.get_tensor_ptr()->get_names());
+
+                            const std::string original_name = op->get_friendly_name();
+                            op->set_friendly_name(original_name + "_original");
+                            convert->set_friendly_name(original_name);
+                        }
+                    }
+                    output.get_tensor_ptr()->set_names({});
+                }
+            }
+        }
+    }
+
+    for (auto it = result_types.begin(); it != result_types.end(); ++it) {
+        const auto result = it->first;
+        const auto actual_type = result->get_input_element_type(0);
+        const auto expected_type = it->second;
+        if (actual_type != it->second) {
+            was_updated = true;
+            auto convert = std::make_shared<ngraph::snippets::op::ConvertSaturation>(
+                result->get_input_node_shared_ptr(0),
+                expected_type);
+            ngraph::copy_runtime_info(result->get_input_node_shared_ptr(0), convert);
+            result->set_argument(0, convert);
+        }
+    }
+
+    return was_updated;
+}
+
+bool ngraph::snippets::pass::PropagatePrecision::can_be_removed(
+    const element::Type& actual_before,
+    const element::Type& actual_after,
+    const element::Type& required_after) noexcept {
+    if (actual_before != required_after) {
+        return false;
+    }
+
+    // CPU support: avoid FP32 => BF16 => FP32
+    if ((actual_before == element::f32) && (actual_after == element::bf16)) {
+        return true;
+    }
+
+    return can_be_fused(actual_after, actual_before);
+}
+
+bool ngraph::snippets::pass::PropagatePrecision::can_be_fused(
+    const element::Type& actual,
+    const element::Type& required) noexcept {
+    // custom conditions: between int & float precisions
+    if (((actual == element::bf16) || (actual == element::f16) || (actual == element::f32)) &&
+        ((required == element::u8) || (required == element::i8))) {
+        return true;
+    }
```

Comment:
What's about `actual = i32`? Honestly I don't understand why you prefer to iterate over all possible cases instead of the general condition.
```suggestion
return required.size() > actual.size() && (actual.is_integer() || actual.is_real() && required.is_real());
```
Maybe I miss something. Could you please explain your choose?

Reply:
Thanks for comment. We should be very accurate here.
1. About `actual = i32`.
   * Case with `actual = i32` is handled at the end of function in general condition.
   * Do you mean `required = i32` : ... => `f32` => `i32` => ... ? In our cases we have only `f32`/`bf16` & `i8` or `f32`/`bf16` & `i32` models . For example: `can_be_removed` function expectes `actual_before` to be equal to `required_after`. In this case (`i32` => `f32` => `i32`), we can not remove `f32`: `f32` mantissa is less then `i32`, and as result we can lost data during `f32` convertion. But in general way: <any precision> => `f32` => `i32`, we can remove `f32` conversion for some cases. Not covered by this PR. New US was created: 105826
2. About generalization. Unfortunatelly, we can not do it. We can not just compare memory for different type representaions (mantisa + exponenta VS mantisa only) by the same reason as described in point 1. Thanks for comment.